net/captivedetection/endpoints.go

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package captivedetection
+
+import (
+	"net/http"
+	"runtime"
+	"slices"
+
+	"tailscale.com/net/dnsfallback"
+)
+
+// availableEndpoints returns a set of Endpoints which can be used for captive portal detection by performing
+// one or more HTTP requests and looking at the response. The returned Endpoints are ordered by preference,
+// with the most preferred Endpoint being the first in the slice.
+func availableEndpoints(preferredDERPRegionID int) []Endpoint {
+	endpoints := []Endpoint{}
+
+	// Before anything, if we have a preferred DERP region, let's try to use that as our first endpoint.
+	preferredIPsAndHosts := []string{}
+	if preferredDERPRegionID != 0 {
+		hs := dnsfallback.GetDERPHostnames(preferredDERPRegionID)
+		ips := dnsfallback.GetDERPIPv4s(preferredDERPRegionID)
+		preferredIPsAndHosts = slices.Concat(ips, hs)
+	}
```

Comment:
This doesn't quite work properly, especially if the user is on a custom DERP server (e.g. Headscale). We might want to just pass a `DERPMap` in here, I think.

Reply:
And that's done (when available).